pkg/runner/runner.go

```diff
@@ -661,17 +719,187 @@ func GetEntryCodeInstructions(function starknet.EntryPointByFunction) ([]*fp.Ele
+				[ap] = %s + 2, ap++;
+				jmp rel 4 if [ap-3] != 0;
+				jmp rel 12;
+
+				[ap] = [[ap-2]], ap++;
+				[ap-1] = [[ap-2]];
+				[ap-4] = [ap] + 1, ap++;
+				[ap] = [ap-4] + 1, ap++;
+				[ap] = [ap-4] + 1, ap++;
+				jmp rel -8 if [ap-3] != 0;
+			`, deref(ApRegister, arrayEndPtr), deref(ApRegister, arrayStartPtr), outputPtr, deref(ApRegister, arrayStartPtr-2), outputPtr),
+		)
+		if paramsSize != 0 {
+			offset := 2*len(programBuiltins) - 1
+			if gotSegmentArena {
+				offset += 4
+			}
+			if gotGasBuiltin {
+				offset += 1
+			}
+			arrayStartPtr := deref(FpRegister, offset)
+			arrayEndPtr := deref(FpRegister, offset+1)
+
+			ctx.AddInlineCASM(
+				fmt.Sprintf(`
+					%s = [ap] + %s, ap++;
+					[ap-1] = [[ap-2]];
+					[ap] = [ap-1], ap++;
+					[ap] = %s, ap++;
+					[ap] = [ap-4]+1, ap++;
+					jmp rel 4 if [ap-3] != 0;
+					jmp rel 12;
+
+					[ap] = [[ap-2]], ap++;
+					[ap-1] = [[ap-2]];
+					[ap-4] = [ap]+1, ap++;
+					[ap] = [ap-4]+1, ap++;
+					[ap] = [ap-4]+1, ap++;
+					jmp rel -8 if [ap-3] != 0;
+				`, arrayEndPtr, arrayStartPtr, arrayStartPtr),
+			)
+		}
+		// After we are done writing into the output segment, we can write the final output_ptr into locals:
+		// The last instruction will write the final output ptr so we can find it in [ap - 1]
+		ctx.AddInlineCASM("[fp] = [ap - 1];")
+
+		if gotSegmentArena {
+			offset := 2 + len(programBuiltins)*2
+			segmentArenaPtr := fmt.Sprintf("[fp + %d]", offset)
+			hints[uint64(ctx.currentCodeOffset)] = append(hints[uint64(ctx.currentCodeOffset)], &core.RelocateAllDictionaries{})
+			ctx.AddInlineCASM(fmt.Sprintf(`
+				[ap]=[%s-2], ap++;
+				[ap]=[%s-1], ap++;
+				[ap-2]=[ap-1];
+				jmp rel 4 if [ap-2] != 0;
+				jmp rel 19;
+				[ap]=[%s-3], ap++;
+				[ap-3] = [ap]+1;
+				jmp rel 4 if [ap-1] != 0;
+				jmp rel 12;
+				[ap]=[[ap-2]+1], ap++;
+				[ap] = [[ap-3]+3], ap++;
+        		[ap-1] = [ap-2] + 1;
+        		[ap] = [ap-4] + 3, ap++;
+        		[ap-4] = [ap] + 1, ap++;
+        		jmp rel -12;
+				`, segmentArenaPtr, segmentArenaPtr, segmentArenaPtr,
+			))
+		}
+
+		// Copying the final builtins from locals into the top of the stack.
+		for i := range programBuiltins {
+			ctx.AddInlineCASM(fmt.Sprintf("[ap] = [fp + %d], ap++;", i))
+		}
+	} else {
+		// Writing the final builtins into the top of the stack.
+		for _, b := range programBuiltins {
+			offset := builtinsOffsetsMap[b]
+			ctx.AddInlineCASM(fmt.Sprintf("[ap] = [fp - %d], ap++;", offset))
+		}
+
+	}
+	if proofmode {
+		ctx.AddInlineCASM("jmp rel 0;")
+	} else {
+		ctx.AddInlineCASM("ret;")
 	}
-	totalSize := uint64(endInstructionsSize) + uint64(codeOffset)
-	ctx.AddInlineCASM(fmt.Sprintf("call rel %d; ret;", int(totalSize)))
-	return ctx.instructions, hints, nil
+	ctx.instructions[callRelArgLocation] = new(fp.Element).SetUint64(uint64(ctx.currentCodeOffset) + codeOffsetBeforeCallRel)
+	return ctx.instructions, hints, ctx.currentCodeOffset, programBuiltins, gotGasBuiltin, gotSegmentArena, nil
```

Comment:
Is this function returning an error ever?

Reply:
no haha, it doesn't. Good catch